File: chundakadan/chundakadan/report/leave_encashment_monthly_balance/leave_encashment_monthly_balance.js
// Copyright (c) 2024, Chundakadan and contributors
// For license information, please see license.txt

frappe.query_reports["Leave Encashment Monthly Balance"] = {
	"filters": [
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 0
		},
		{
			"fieldname": "employee",
			"label": __("Employee"),
			"fieldtype": "Link",
			"options": "Employee",
			"reqd": 1,
			"on_change": function() {
				let employee = frappe.query_report.get_filter_value('employee');
				if (employee) {
					frappe.db.get_value('Employee', employee, 'relieving_date', (r) => {
						if (r && r.relieving_date) {
							frappe.query_report.set_filter_value('to_date', r.relieving_date);
						} else {
							frappe.query_report.set_filter_value('to_date', frappe.datetime.get_today());
						}
					});
				}
			}
		},
		{
			"fieldname": "leave_period",
			"label": __("Leave Period"),
			"fieldtype": "Link",
			"options": "Leave Period",
			"reqd": 0
		},
		{
			"fieldname": "to_date",
			"label": __("Reference Date (Relieving Date)"),
			"fieldtype": "Date",
			"reqd": 0
		},
		{
			"fieldname": "department",
			"label": __("Department"),
			"fieldtype": "Link",
			"options": "Department",
			"reqd": 0
		},
		{
			"fieldname": "leave_type",
			"label": __("Leave Type"),
			"fieldtype": "Link",
			"options": "Leave Type",
			"get_query": function() {
				return {
					"filters": {
						"allow_encashment": 1
					}
				};
			},
			"reqd": 0
		},
		{
			"fieldname": "payment_days",
			"label": __("Payment Days"),
			"fieldtype": "Int",
			"default": 30,
			"reqd": 1
		},
		{
			"fieldname": "salary_component",
			"label": __("Leave Encashment Salary Component"),
			"fieldtype": "Link",
			"options": "Salary Component",
			"get_query": function() {
				return {
					"filters": {
						"type": "Earning"
					}
				};
			},
			"reqd": 0
		}
	],
	
	"onload": function(report) {
		// Add bulk create button
		report.page.add_inner_button(__("Bulk Create Additional Salary"), function() {
			bulk_create_additional_salary(report);
		});
		
		// Add selective create button
		report.page.add_inner_button(__("Create Individual Additional Salary"), function() {
			selective_create_additional_salary(report);
		});
	},
	
	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		if (column.fieldname == "pending_encashment_leaves" && data && data.pending_encashment_leaves > 0) {
			value = "<span style='color:orange; font-weight:bold'>" + value + "</span>";
		}
		
		if (column.fieldname == "total_payable_amount" && data && data.total_payable_amount > 0) {
			value = "<span style='color:green; font-weight:bold'>" + value + "</span>";
		}
		
		return value;
	}
};

// Bulk create Additional Salary for all employees
function bulk_create_additional_salary(report) {
	// Get salary component from filters
	let salary_component = frappe.query_report.get_filter_value('salary_component');
	
	if (!salary_component) {
		frappe.msgprint({
			title: __('Salary Component Required'),
			message: __('Please select a Leave Encashment Salary Component in the filters before creating Additional Salary'),
			indicator: 'red'
		});
		return;
	}
	
	// Get report data
	let data = frappe.query_report.data;
	
	if (!data || data.length === 0) {
		frappe.msgprint(__('No data available. Please run the report first.'));
		return;
	}
	
	// Filter employees with pending encashment
	let eligible_employees = data.filter(row => 
		row.pending_encashment_leaves > 0 && 
		row.total_payable_amount > 0 &&
		row.employee &&
		!row.employee.toLowerCase().includes('total')
	);
	
	if (eligible_employees.length === 0) {
		frappe.msgprint(__('No employees with pending leave encashment found.'));
		return;
	}
	
	// Calculate totals
	let employee_totals = {};
	eligible_employees.forEach(row => {
		if (!row.employee || typeof row.employee !== 'string') return;
		
		let key = row.employee + '|' + row.leave_type;
		if (!employee_totals[key]) {
			employee_totals[key] = {
				employee: row.employee,
				employee_name: row.employee_name || row.employee,
				leave_type: row.leave_type,
				pending_leaves: row.pending_encashment_leaves,
				amount: row.total_payable_amount
			};
		}
	});
	
	let unique_employees = Object.values(employee_totals);
	let total_employees = new Set(unique_employees.map(e => e.employee)).size;
	let total_amount = unique_employees.reduce((sum, e) => sum + e.amount, 0);
	let total_leaves = unique_employees.reduce((sum, e) => sum + e.pending_leaves, 0);
	
	// Confirmation UI
	let breakdown_html = Object.keys(employee_totals).map(key => {
		let e = employee_totals[key];
		return `<li>${e.employee_name} (${e.employee}): ${e.leave_type}</li>`;
	}).join('');
	
	let message = `
		<div style="margin-bottom: 15px;">
			<h4>Bulk Create Additional Salary</h4>
			<p>This will create Additional Salary entries for all eligible employees.</p>
		</div>
		<table class="table table-bordered" style="margin-bottom: 15px;">
			<tr><td><strong>Total Employees:</strong></td><td>${total_employees}</td></tr>
			<tr><td><strong>Total Entries:</strong></td><td>${unique_employees.length}</td></tr>
			<tr><td><strong>Total Leaves:</strong></td><td>${total_leaves.toFixed(2)}</td></tr>
			<tr><td><strong>Total Amount:</strong></td><td>${format_currency(total_amount)}</td></tr>
			<tr><td><strong>Salary Component:</strong></td><td>${salary_component}</td></tr>
		</table>
		<div style="margin-bottom: 15px;">
			<strong>Entries:</strong>
			<ul style="max-height: 200px; overflow-y: auto; margin-top: 5px;">${breakdown_html}</ul>
		</div>
	`;
	
	frappe.confirm(message, function() {
		frappe.prompt([
			{
				'fieldname': 'payroll_date',
				'fieldtype': 'Date',
				'label': __('Payroll Date'),
				'reqd': 1,
				'default': frappe.datetime.get_today()
			}
		], function(values) {
			let employees_data = unique_employees.map(e => ({
				employee: e.employee,
				employee_name: e.employee_name,
				leave_type: e.leave_type,
				pending_leaves: e.pending_leaves,
				amount: e.amount,
				company: eligible_employees.find(row => row.employee === e.employee && row.leave_type === e.leave_type).company || ''
			}));
			
			frappe.call({
				method: 'chundakadan.chundakadan.report.leave_encashment_monthly_balance.leave_encashment_monthly_balance.bulk_create_additional_salary',
				args: {
					employees_data: employees_data,
					salary_component: salary_component,
					payroll_date: values.payroll_date,
					overwrite_salary_structure_amount: 0
				},
				freeze: true,
				freeze_message: __('Creating Additional Salary entries...'),
				callback: function(r) {
					if (r.message) {
						let result = r.message;
						let result_message = `
							<h4>Bulk Creation Results</h4>
							<table class="table table-bordered">
								<tr class="success"><td><strong>Created:</strong></td><td>${result.created}</td></tr>
								<tr class="warning"><td><strong>Skipped:</strong></td><td>${result.skipped}</td></tr>
								<tr class="danger"><td><strong>Failed:</strong></td><td>${result.failed}</td></tr>
							</table>
						`;
						frappe.msgprint({ title: __('Complete'), message: result_message, wide: true });
						frappe.query_report.refresh();
					}
				}
			});
		});
	});
}

// Selective create Additional Salary
function selective_create_additional_salary(report) {
	let salary_component = frappe.query_report.get_filter_value('salary_component');
	if (!salary_component) {
		frappe.msgprint(__('Please select a Salary Component'));
		return;
	}
	
	let data = frappe.query_report.data;
	let eligible_employees = data.filter(row => 
		row.pending_encashment_leaves > 0 && row.total_payable_amount > 0 && row.employee && !row.employee.toLowerCase().includes('total')
	);
	
	if (eligible_employees.length === 0) {
		frappe.msgprint(__('No eligible employees found'));
		return;
	}
	
	let employee_options = {};
	eligible_employees.forEach(row => {
		if (!employee_options[row.employee]) {
			employee_options[row.employee] = {
				employee: row.employee,
				employee_name: row.employee_name,
				company: row.company || '',
				leave_types: []
			};
		}
		employee_options[row.employee].leave_types.push(row);
	});
	
	show_selection_dialog(employee_options, salary_component);
}

function show_selection_dialog(employee_options, salary_component) {
	let employees_list = Object.values(employee_options);
	let html = `
		<div style="max-height: 500px; overflow-y: auto;">
			<div id="selection-summary" style="background: #eee; padding: 10px; margin-bottom: 10px; border-radius: 4px;">
				Selected: <span id="selected-count">0</span> | Amount: <span id="selected-amount">0</span>
			</div>
			<div class="list-group">
	`;
	
	employees_list.forEach((emp, e_idx) => {
		html += `<div class="list-group-item">
			<h5 class="mb-1"><input type="checkbox" class="employee-checkbox" data-employee="${emp.employee}"> ${emp.employee_name}</h5>
			<div style="margin-left: 20px;">
		`;
		emp.leave_types.forEach((lt, l_idx) => {
			html += `
				<div class="form-check">
					<input class="form-check-input leave-type-checkbox" type="checkbox" 
						data-employee="${emp.employee}" data-leave-type="${lt.leave_type}" 
						data-amount="${lt.total_payable_amount}" data-leaves="${lt.pending_encashment_leaves}"
						data-company="${emp.company}" data-name="${emp.employee_name}">
					<label class="form-check-label">${lt.leave_type} (${lt.pending_encashment_leaves} days @ ${format_currency(lt.total_payable_amount)})</label>
				</div>
			`;
		});
		html += `</div></div>`;
	});
	html += `</div></div>`;

	let d = new frappe.ui.Dialog({
		title: __('Select for Additional Salary'),
		fields: [
			{ fieldtype: 'HTML', fieldname: 'selection_html', options: html },
			{ fieldtype: 'Date', fieldname: 'payroll_date', label: __('Payroll Date'), reqd: 1, default: frappe.datetime.get_today() }
		],
		size: 'large',
		primary_action_label: __('Create Selected'),
		primary_action: function(values) {
			let selected = [];
			d.$wrapper.find('.leave-type-checkbox:checked').each(function() {
				let $cb = $(this);
				selected.push({
					employee: $cb.data('employee'),
					employee_name: $cb.data('name'),
					company: $cb.data('company'),
					leave_type: $cb.data('leave-type'),
					pending_leaves: $cb.data('leaves'),
					amount: $cb.data('amount')
				});
			});
			
			if (selected.length === 0) return;
			
			frappe.call({
				method: 'chundakadan.chundakadan.report.leave_encashment_monthly_balance.leave_encashment_monthly_balance.bulk_create_additional_salary',
				args: {
					employees_data: selected,
					salary_component: salary_component,
					payroll_date: values.payroll_date
				},
				freeze: true,
				callback: function(r) {
					if (r.message) {
						frappe.msgprint(__('Created {0} entries', [r.message.created]));
						d.hide();
						frappe.query_report.refresh();
					}
				}
			});
		}
	});

	d.show();
	
	d.$wrapper.find('.employee-checkbox').on('change', function() {
		let emp = $(this).data('employee');
		d.$wrapper.find(`.leave-type-checkbox[data-employee="${emp}"]`).prop('checked', $(this).prop('checked'));
		update_summary();
	});

	d.$wrapper.find('.leave-type-checkbox').on('change', function() {
		update_summary();
	});

	function update_summary() {
		let count = 0;
		let amount = 0;
		d.$wrapper.find('.leave-type-checkbox:checked').each(function() {
			count++;
			amount += parseFloat($(this).data('amount'));
		});
		d.$wrapper.find('#selected-count').text(count);
		d.$wrapper.find('#selected-amount').text(format_currency(amount));
	}
}

File: chundakadan/chundakadan/report/leave_encashment_monthly_balance/leave_encashment_monthly_balance.py
# Copyright (c) 2024, Chundakadan and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, getdate, add_months, get_last_day
from hrms.hr.doctype.leave_application.leave_application import get_leave_balance_on

def execute(filters=None):
    if not filters:
        filters = {}
    
    columns = get_columns(filters)
    data = get_data(filters)
    chart = get_chart_data(data)
    summary = get_report_summary(data)
    
    return columns, data, None, chart, summary

def get_columns(filters):
    return [
        {
            "label": _("Employee"),
            "fieldname": "employee",
            "fieldtype": "Link",
            "options": "Employee",
            "width": 120
        },
        {
            "label": _("Employee Name"),
            "fieldname": "employee_name",
            "fieldtype": "Data",
            "width": 180
        },
        {
            "label": _("Department"),
            "fieldname": "department",
            "fieldtype": "Link",
            "options": "Department",
            "width": 130
        },
        {
            "label": _("Leave Type"),
            "fieldname": "leave_type",
            "fieldtype": "Link",
            "options": "Leave Type",
            "width": 130
        },
        {
            "label": _("Encashment Enabled"),
            "fieldname": "encashment_enabled",
            "fieldtype": "Check",
            "width": 100
        },
        {
            "label": _("Total Eligible Leaves"),
            "fieldname": "total_eligible_leaves",
            "fieldtype": "Float",
            "width": 110,
            "precision": 2
        },
        {
            "label": _("Leaves Already Encashed"),
            "fieldname": "leaves_already_encashed",
            "fieldtype": "Float",
            "width": 120,
            "precision": 2
        },
        {
            "label": _("Leaves Taken"),
            "fieldname": "leaves_taken",
            "fieldtype": "Float",
            "width": 100,
            "precision": 2
        },
        {
            "label": _("Current Balance"),
            "fieldname": "current_balance",
            "fieldtype": "Float",
            "width": 110,
            "precision": 2
        },
        {
            "label": _("Pending for Encashment"),
            "fieldname": "pending_encashment_leaves",
            "fieldtype": "Float",
            "width": 130,
            "precision": 2
        },
        {
            "label": _("Basic Salary"),
            "fieldname": "basic_salary",
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "label": _("Payment Days"),
            "fieldname": "payment_days",
            "fieldtype": "Int",
            "width": 90
        },
        {
            "label": _("Per Day Rate"),
            "fieldname": "per_day_rate",
            "fieldtype": "Currency",
            "width": 110
        },
        {
            "label": _("Total Payable Amount"),
            "fieldname": "total_payable_amount",
            "fieldtype": "Currency",
            "width": 140
        }
    ]

def get_data(filters):
    company = filters.get("company")
    employee_id = filters.get("employee")
    leave_type_filter = filters.get("leave_type")
    payment_days = flt(filters.get("payment_days", 30))
    department = filters.get("department")

    if not employee_id:
        frappe.msgprint(_("Please select an Employee"))
        return []

    # Get employee details (joining and relieving date)
    employee_details = frappe.db.get_value("Employee", employee_id, ["date_of_joining", "relieving_date", "employee_name", "department", "company"], as_dict=True)
    if not employee_details:
        frappe.msgprint(_("Employee not found"))
        return []

    from_date = employee_details.date_of_joining
    relieving_date = employee_details.relieving_date

    if not from_date:
        frappe.msgprint(_("Employee Joining Date is missing"))
        return []

    # The calculation window should always end at the last day of the month 
    # BEFORE the reference/relieving date provided.
    ref_date = filters.get("to_date")
    if not ref_date:
        ref_date = relieving_date or frappe.utils.today()
    
    to_date = get_last_day(add_months(getdate(ref_date), -1))

    # Get Encashable Leave Types
    lt_filters = {"allow_encashment": 1}
    if leave_type_filter:
        lt_filters["name"] = leave_type_filter
        
    encashable_leave_types = frappe.get_all(
        "Leave Type", 
        filters=lt_filters, 
        fields=["name", "max_encashable_leaves", "non_encashable_leaves"]
    )
    encashable_lt_map = {lt.name: lt for lt in encashable_leave_types}
    
    if not encashable_lt_map:
        frappe.msgprint(_("No encashment-enabled leave types found"))
        return []

    # Fetch Leave Allocations for the employee in the joined period
    allocation_query = """
        SELECT 
            name, employee, employee_name, department, leave_type,
            total_leaves_allocated, total_leaves_encashed, from_date, to_date
        FROM `tabLeave Allocation`
        WHERE docstatus = 1
        AND employee = %(employee)s
        AND leave_type IN %(leave_types)s
        AND (from_date <= %(to_date)s AND to_date >= %(from_date)s)
    """
    
    allocations = frappe.db.sql(allocation_query, {
        "employee": employee_id,
        "leave_types": tuple(encashable_lt_map.keys()),
        "from_date": from_date,
        "to_date": to_date
    }, as_dict=True)

    if not allocations:
        # If no allocations found in the period, check if there are ANY allocations to avoid silent failure
        return []

    # Fetch Salary Structure Assignment
    salary_map = get_basic_salaries([employee_id])
    
    # Get leaves taken in the period
    leaves_taken_map = get_leaves_taken([employee_id], list(encashable_lt_map.keys()), from_date, to_date)

    data = []
    
    for alloc in allocations:
        lt_settings = encashable_lt_map.get(alloc.leave_type)
        basic_salary = flt(salary_map.get(alloc.employee, 0))
        
        # Prorate calculation (Month Based)
        # Calculation: (Allocated Leaves / Total Months in Allocation) * Months Worked (up to to_date)
        
        alloc_from = getdate(alloc.from_date)
        alloc_to = getdate(alloc.to_date)
        
        # Total months in the allocation period
        total_months = frappe.utils.month_diff(alloc_to, alloc_from)
        
        # Months elapsed until the end of previous month
        if to_date >= alloc_from:
            worked_months = frappe.utils.month_diff(to_date, alloc_from)
        else:
            worked_months = 0
            
        if total_months > 0:
            prorated_allocation = (flt(alloc.total_leaves_allocated) / total_months) * worked_months
        else:
            prorated_allocation = 0
            
        # Remove "pointed values" (decimals) as requested
        prorated_allocation = int(prorated_allocation)
        
        # Get leaves taken till to_date
        leaves_taken = flt(leaves_taken_map.get((alloc.employee, alloc.leave_type), 0))
        
        # Leaves already encashed from this allocation
        leaves_encashed = flt(alloc.total_leaves_encashed)
        
        # Calculate Balance on the calculated to_date (Prorated)
        # Current Balance = Prorated Allocation - Taken - Encashed
        balance = max(0, prorated_allocation - leaves_taken - leaves_encashed)
        
        # Calculate Pending Encashable Leaves
        max_encashable = flt(lt_settings.max_encashable_leaves) if lt_settings.max_encashable_leaves else 999999
        non_encashable = flt(lt_settings.non_encashable_leaves)
        
        encashable_balance = max(0, balance - non_encashable)
        pending_leaves = min(encashable_balance, max_encashable)
        
        # Calculate Amount
        per_day_rate = 0
        if payment_days > 0:
            per_day_rate = basic_salary / payment_days
            
        payable_amount = pending_leaves * per_day_rate
        
        row = {
            "employee": alloc.employee,
            "employee_name": alloc.employee_name,
            "department": alloc.department,
            "leave_type": alloc.leave_type,
            "company": employee_details.company or "",
            "encashment_enabled": 1,
            "total_eligible_leaves": prorated_allocation,
            "leaves_already_encashed": flt(alloc.total_leaves_encashed, 2),
            "leaves_taken": flt(leaves_taken, 2),
            "current_balance": flt(balance, 2),
            "pending_encashment_leaves": flt(pending_leaves, 2),
            "basic_salary": flt(basic_salary, 2),
            "payment_days": int(payment_days),
            "per_day_rate": flt(per_day_rate, 2),
            "total_payable_amount": flt(payable_amount, 2)
        }
        
        data.append(row)

    return data

def get_basic_salaries(employees):
    """Fetch the latest active salary structure assignment for each employee"""
    if not employees:
        return {}
    salary_map = {}
    assignments = frappe.db.sql("""
        SELECT employee, base
        FROM `tabSalary Structure Assignment`
        WHERE employee IN %(employees)s
        AND docstatus = 1
        ORDER BY from_date DESC
    """, {"employees": tuple(employees)}, as_dict=True)
    seen_employees = set()
    for assign in assignments:
        if assign.employee not in seen_employees:
            salary_map[assign.employee] = flt(assign.base)
            seen_employees.add(assign.employee)
    return salary_map

def get_leaves_taken(employees, leave_types, from_date, to_date):
    """Get total leaves taken by each employee for each leave type in the period"""
    if not employees or not leave_types:
        return {}
    leaves_taken = frappe.db.sql("""
        SELECT employee, leave_type, SUM(total_leave_days) as total_taken
        FROM `tabLeave Application`
        WHERE employee IN %(employees)s
        AND leave_type IN %(leave_types)s
        AND docstatus = 1
        AND status = 'Approved'
        AND from_date >= %(from_date)s
        AND to_date <= %(to_date)s
        GROUP BY employee, leave_type
    """, {
        "employees": tuple(employees),
        "leave_types": tuple(leave_types),
        "from_date": from_date,
        "to_date": to_date
    }, as_dict=True)
    leaves_map = {}
    for record in leaves_taken:
        leaves_map[(record.employee, record.leave_type)] = flt(record.total_taken)
    return leaves_map

def get_report_summary(data):
    if not data:
        return []
    total_employees = len(set(row["employee"] for row in data))
    total_pending_leaves = sum(flt(row["pending_encashment_leaves"]) for row in data)
    total_encashed_leaves = sum(flt(row["leaves_already_encashed"]) for row in data)
    total_payable = sum(flt(row["total_payable_amount"]) for row in data)
    return [
        {"value": total_employees, "label": _("Total Employees"), "datatype": "Int", "indicator": "blue"},
        {"value": total_pending_leaves, "label": _("Total Pending Leaves"), "datatype": "Float", "indicator": "orange"},
        {"value": total_encashed_leaves, "label": _("Total Encashed Leaves"), "datatype": "Float", "indicator": "green"},
        {"value": total_payable, "label": _("Total Payable Amount"), "datatype": "Currency", "indicator": "blue"}
    ]

def get_chart_data(data):
    if not data:
        return None
    leave_type_data = {}
    for row in data:
        lt = row["leave_type"]
        if lt not in leave_type_data:
            leave_type_data[lt] = {"pending": 0, "encashed": 0, "payable": 0}
        leave_type_data[lt]["pending"] += flt(row["pending_encashment_leaves"])
        leave_type_data[lt]["encashed"] += flt(row["leaves_already_encashed"])
        leave_type_data[lt]["payable"] += flt(row["total_payable_amount"])
    labels = list(leave_type_data.keys())
    pending_values = [leave_type_data[lt]["pending"] for lt in labels]
    encashed_values = [leave_type_data[lt]["encashed"] for lt in labels]
    return {
        "data": {
            "labels": labels,
            "datasets": [
                {"name": _("Pending for Encashment"), "values": pending_values},
                {"name": _("Already Encashed"), "values": encashed_values}
            ]
        },
        "type": "bar",
        "colors": ["#FFA500", "#28a745"]
    }

@frappe.whitelist()
def create_additional_salary(employee, leave_type, pending_leaves, amount, salary_component, 
                            payroll_date, overwrite_salary_structure_amount=1, company=None):
    from frappe.utils import getdate
    if not employee or not salary_component or not amount:
        frappe.throw(_("Employee, Salary Component, and Amount are required"))
    amount = flt(amount)
    pending_leaves = flt(pending_leaves)
    if amount <= 0: frappe.throw(_("Amount must be greater than zero"))
    if pending_leaves <= 0: frappe.throw(_("Pending leaves must be greater than zero"))
    employee_doc = frappe.get_doc("Employee", employee)
    if not company: company = employee_doc.company
    existing = frappe.db.exists("Additional Salary", {
        "employee": employee, "salary_component": salary_component, "payroll_date": getdate(payroll_date),
        "ref_doctype": "Leave Type", "ref_docname": leave_type, "docstatus": ["<", 2]
    })
    if existing: return existing
    additional_salary = frappe.get_doc({
        "doctype": "Additional Salary", "employee": employee, "employee_name": employee_doc.employee_name,
        "company": company, "salary_component": salary_component, "amount": amount,
        "payroll_date": getdate(payroll_date), "overwrite_salary_structure_amount": int(overwrite_salary_structure_amount),
        "ref_doctype": "Leave Type", "ref_docname": leave_type,
        "remarks": _("Monthly Leave Encashment: {0} leaves").format(pending_leaves)
    })
    additional_salary.insert()
    return additional_salary.name

@frappe.whitelist()
def bulk_create_additional_salary(employees_data, salary_component, payroll_date, 
                                  overwrite_salary_structure_amount=1):
    import json
    if isinstance(employees_data, str): employees_data = json.loads(employees_data)
    results = {"total": len(employees_data), "created": 0, "skipped": 0, "failed": 0, "errors": [], "created_docs": []}
    payroll_date = get_last_day(getdate(payroll_date))
    for emp_data in employees_data:
        try:
            employee = emp_data.get("employee")
            amount = flt(emp_data.get("amount", 0))
            if not employee or amount <= 0: continue
            employee_doc = frappe.get_doc("Employee", employee)
            existing = frappe.db.exists("Additional Salary", {
                "employee": employee, "salary_component": salary_component, "payroll_date": payroll_date,
                "ref_doctype": "Leave Type", "ref_docname": emp_data.get("leave_type"), "docstatus": ["<", 2]
            })
            if existing:
                results["skipped"] += 1
                continue
            additional_salary = frappe.get_doc({
                "doctype": "Additional Salary", "employee": employee, "employee_name": employee_doc.employee_name,
                "company": employee_doc.company, "salary_component": salary_component, "amount": amount,
                "payroll_date": payroll_date, "overwrite_salary_structure_amount": int(overwrite_salary_structure_amount),
                "ref_doctype": "Leave Type", "ref_docname": emp_data.get("leave_type"),
                "remarks": _("Monthly Leave Encashment: {0} leaves").format(emp_data.get("pending_leaves"))
            })
            additional_salary.insert()
            results["created"] += 1
            results["created_docs"].append(additional_salary.name)
        except Exception as e:
            results["failed"] += 1
            results["errors"].append(str(e))
    frappe.db.commit()
    return results
